### PR TITLE
[subnet decap] Add subnet decap rule based on overlay ECMP

### PR DESCRIPTION
**Pull request by lolyu**
**What I did**
Support dynamic decap rule generation based on the Vxlan tunnel route of Overlay ECMP.
This depends on: https://github.com/sonic-net/sonic-swss/pull/3117

**Why I did it**
To enable SONiC with the capability to decap IPinIP packets with dest IP in the Overlay ECMP Vxlan tunnel route prefix.

**How I verified it**
VS test:
```
test_vnet.py::TestVnetOrch::test_vnet_orch_1 PASSED                                                                                                                                            [  3%]
test_vnet.py::TestVnetOrch::test_vnet_orch_2 PASSED                                                                                                                                            [  6%]
test_vnet.py::TestVnetOrch::test_vnet_orch_3 PASSED                                                                                                                                            [  9%]
test_vnet.py::TestVnetOrch::test_vnet_orch_4 SKIPPED (Failing. Under investigation)                                                                                                            [ 12%]
test_vnet.py::TestVnetOrch::test_vnet_orch_5 PASSED                                                                                                                                            [ 15%]
test_vnet.py::TestVnetOrch::test_vnet_vxlan_multi_map PASSED                                                                                                                                   [ 18%]
test_vnet.py::TestVnetOrch::test_vnet_orch_7[true] PASSED                                                                                                                                      [ 21%]
test_vnet.py::TestVnetOrch::test_vnet_orch_7[false] PASSED                                                                                                                                     [ 24%]
test_vnet.py::TestVnetOrch::test_vnet_orch_8[true] PASSED                                                                                                                                      [ 27%]
test_vnet.py::TestVnetOrch::test_vnet_orch_8[false] PASSED                                                                                                                                     [ 30%]
test_vnet.py::TestVnetOrch::test_vnet_orch_9[true] PASSED                                                                                                                                      [ 33%]
test_vnet.py::TestVnetOrch::test_vnet_orch_9[false] PASSED                                                                                                                                     [ 36%]
test_vnet.py::TestVnetOrch::test_vnet_orch_10[true] PASSED                                                                                                                                     [ 39%]
test_vnet.py::TestVnetOrch::test_vnet_orch_10[false] PASSED                                                                                                                                    [ 42%]
test_vnet.py::TestVnetOrch::test_vnet_orch_11[true] PASSED                                                                                                                                     [ 45%]
test_vnet.py::TestVnetOrch::test_vnet_orch_11[false] PASSED                                                                                                                                    [ 48%]
test_vnet.py::TestVnetOrch::test_vnet_orch_12 PASSED                                                                                                                                           [ 51%]
test_vnet.py::TestVnetOrch::test_vnet_orch_13 PASSED                                                                                                                                           [ 54%]
test_vnet.py::TestVnetOrch::test_vnet_orch_14 PASSED                                                                                                                                           [ 57%]
test_vnet.py::TestVnetOrch::test_vnet_orch_15 PASSED                                                                                                                                           [ 60%]
test_vnet.py::TestVnetOrch::test_vnet_orch_16 PASSED                                                                                                                                           [ 63%]
test_vnet.py::TestVnetOrch::test_vnet_orch_17 PASSED                                                                                                                                           [ 66%]
test_vnet.py::TestVnetOrch::test_vnet_orch_18 PASSED                                                                                                                                           [ 69%]
test_vnet.py::TestVnetOrch::test_vnet_orch_19 PASSED                                                                                                                                           [ 72%]
test_vnet.py::TestVnetOrch::test_vnet_orch_20 PASSED                                                                                                                                           [ 75%]
test_vnet.py::TestVnetOrch::test_vnet_orch_21 PASSED                                                                                                                                           [ 78%]
test_vnet.py::TestVnetOrch::test_vnet_orch_22 PASSED                                                                                                                                           [ 81%]
test_vnet.py::TestVnetOrch::test_vnet_orch_23 PASSED                                                                                                                                           [ 84%]
test_vnet.py::TestVnetOrch::test_vnet_orch_24 PASSED                                                                                                                                           [ 87%]
test_vnet.py::TestVnetOrch::test_vnet_orch_25 PASSED                                                                                                                                           [ 90%]
test_vnet.py::TestVnetOrch::test_vnet_orch_26 PASSED                                                                                                                                           [ 93%]
test_vnet.py::TestVnetOrch::test_vnet_orch_27 PASSED                                                                                                                                           [ 96%]
test_vnet.py::test_nonflaky_dummy PASSED                                                                                                                                                       [100%]
```

**Details if related**

### CODE REVIEW

File: orchagent/vnetorch.cpp [hunk 1]
```diff
@@ -2088,6 +2123,14 @@ void VNetRouteOrch::postRouteState(const string& vnet, IpPrefix& ipPrefix, NextH
             removeRouteAdvertisement(prefix_to_use);
         }
     }
+    if (route_state == "active")
```
Comment (bingwang-ms) [inline orchagent/vnetorch.cpp:2126]: Should this code be in the if block at line#2115?

File: orchagent/vnetorch.cpp [hunk 2]
```diff
@@ -1432,6 +1435,38 @@ bool VNetRouteOrch::updateTunnelRoute(const string& vnet, IpPrefix& ipPrefix,
...
+    string tunnel_name = ipPrefix.isV4() ? config.tunnel : config.tunnel_v6;
+    string key = tunnel_name + ":" + ipPrefix.to_string();
+    app_tunnel_decap_term_producer_.set(key, data);
+    subnet_decap_terms_created_.insert(ipPrefix);
+}
+
+inline void VNetRouteOrch::removeSubnetDecapTerm(const IpPrefix &ipPrefix)
+{
+    const SubnetDecapConfig &config = gTunneldecapOrch->getSubnetDecapConfig();
+    if (!config.enable || subnet_decap_terms_created_.find(ipPrefix) == subnet_decap_terms_created_.end())
```
Comment (bingwang-ms) [inline orchagent/vnetorch.cpp:1459]: Suggest not checking `config.enable`. Even if it's disabled, user should be able to remove it.

File: orchagent/vnetorch.cpp (hunk 1 above)
Comment (lolyu) [reply orchagent/vnetorch.cpp:2126]: The if block at LINE#2115 checks if the vnet is allowed to advertise prefix. The decap term should be added even if the  vnet is not allowed to advertise prefix.

**Review by bingwang-ms — APPROVED** [review]

**Comment by prsunny** [comment]
@lolyu , we need to ensure test_vnet_vxlan in sonic-mgmt passes and no stale decap entries present. Could you please confirm?

**Comment by lolyu** [comment]
> @lolyu , we need to ensure test_vnet_vxlan in sonic-mgmt passes and no stale decap entries present. Could you please confirm?

Hi @prsunny, as long as subnet decap feature is not enabled, no more decap term entries will be created. So `test_vnet_vxlan` should be safe here.

**Review by prsunny — APPROVED** [review]

**Pull request merged**